src/hotspot/cpu/s390/templateTable_s390.cpp

```diff
@@ -76,10 +77,13 @@
   __ bind(lbl);                                                                \
   { unsigned int b_off = __ offset();                                          \
     uintptr_t   b_addr = (uintptr_t)__ pc();                                   \
-    __ z_larl(Z_R0, (int64_t)0);     /* Check current address alignment. */    \
-    __ z_slgr(Z_R0, br_tab);         /* Current Address must be equal    */    \
-    __ z_slgr(Z_R0, flags);          /* to calculated branch target.     */    \
-    __ z_brc(Assembler::bcondLogZero, 3); /* skip trap if ok. */               \
+    __ z_larl(br_tab_temp, (int64_t)0);  /* Check current address alignment. */\
+    __ z_slgr(br_tab_temp, br_tab);      /* Current Address must be equal    */\
+    if (tos_state != Z_R0_scratch) {                                           \
+      __ z_slgr(br_tab_temp, tos_state); /* to calculated branch target.     */\
+    }                                                                          \
+    __ z_brc(Assembler::bcondLogZero, 4);/* skip trap if ok. */                \
+    __ z_illtrap(0x55);                                                        \
```

Comment:
Is z_illtrap duplicated by intention? If so, please explain in a comment.

Reply:
The duplication is to circumvent a disassembler shortcoming. It does not honor the instruction length for unknown (illegal) instructions. One of them should be removed now.